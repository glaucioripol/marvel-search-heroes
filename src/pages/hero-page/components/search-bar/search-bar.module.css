.search {
  display: flex;
  justify-content: center;
  width: 100%;
}

.search__form {
  align-items: center;
  background-color: #fff;
  border: transparent;
  border-radius: 32px;
  display: flex;
  justify-content: center;
  height: 44px;
  width: 100%;
  max-width: 620px;
}

.search .search__input {
  background: transparent;
  border: transparent;
  border-radius: 0 32px 32px 0;
  font-size: var(--font-xx-sm);
  height: 100%;
  padding: 24px;
  width: 100%;

  color: var(--red-5);
}

.search .search__input::placeholder {
  color: var(--grey-4);
}

.search .search__input:focus {
  outline: 1px solid var(--red-5);
}

.search .search__button {
  background-color: transparent;
  border: none;
  cursor: pointer;
  height: 36px;
  margin-inline-start: 12px;
  width: 36px;
}
