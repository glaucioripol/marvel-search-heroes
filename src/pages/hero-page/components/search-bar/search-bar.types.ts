export type SearchBarProperties = Readonly<{
  handleSearch?: (value: string) => void;
  disabled?: boolean;
}>;
