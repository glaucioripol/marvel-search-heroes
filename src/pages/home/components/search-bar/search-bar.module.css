.search {
  display: flex;
  justify-content: center;
  margin-block-start: 48px;
}

.search .search__form {
  align-items: center;
  background-color: var(--red-1);
  border: transparent;
  border-radius: 32px;
  display: flex;
  justify-content: center;
  height: 66px;
  width: 100%;
  max-width: 866px;
}

.search .search__input {
  background: transparent;
  border: transparent;
  border-radius: 0 32px 32px 0;
  font-size: var(--font-x-sm);
  height: 100%;
  padding: 24px;
  width: 100%;

  color: var(--red-4);
}

.search .search__input::placeholder {
  color: var(--red-3);
}

.search .search__input:focus {
  outline: 2px solid var(--red-5);
}

.search .search__button {
  background-color: transparent;
  border: none;
  cursor: pointer;
  height: 66px;
  margin-inline-start: 12px;
  width: 66px;
}
