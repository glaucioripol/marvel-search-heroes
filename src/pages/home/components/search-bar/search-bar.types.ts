import { FormEvent } from "react";

export type SearchBarProperties = Readonly<{
  handleSearch: (event: FormEvent<HTMLFormElement>) => void;
  disabled?: boolean;
  inputValue?: string;
}>;
