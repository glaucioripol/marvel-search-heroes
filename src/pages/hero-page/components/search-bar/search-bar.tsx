import styles from "./search-bar.module.css";

import { useCallback } from "react";

import { SearchBarProperties } from "./search-bar.types";

export function SearchBar({ handleSearch, disabled }: SearchBarProperties) {
  const handleOnSearch = useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      const search = new FormData(event.currentTarget);
      const value = search.get("search-hero") as string;

      handleSearch?.(value);
    },
    [handleSearch],
  );

  return (
    <div className={styles.search} data-testid="search-bar-wrapper">
      <form
        className={styles.search__form}
        data-testid="search-form"
        onSubmit={handleOnSearch}>
        <button
          data-testid="search-button"
          className={styles.search__button}
          type="submit"
          aria-label="Buscar por heróis"
          title="Buscar por heróis"
          disabled={disabled}>
          <img
            src="/assets/ic_busca.svg"
            alt="Ícone de uma lupa representando a busca."
            width={22}
            height={22}
          />
        </button>

        <input
          className={styles.search__input}
          type="search"
          aria-label="Procure por heróis"
          title="Procure por heróis"
          placeholder="Procure por heróis (Eu funciono de verdade!)"
          data-testid="search-input"
          name="search-hero"
          maxLength={120}
          disabled={disabled}
          required
        />
      </form>
    </div>
  );
}
