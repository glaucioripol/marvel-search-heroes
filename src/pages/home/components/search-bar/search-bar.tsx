import styles from "./search-bar.module.css";

import { SearchBarProperties } from "./search-bar.types";

export function SearchBar({
  disabled,
  handleSearch,
  inputValue,
}: SearchBarProperties) {
  return (
    <div className={styles.search} data-testid="search-bar-wrapper">
      <form
        className={styles.search__form}
        data-testid="search-form"
        onSubmit={handleSearch}>
        <button
          data-testid="search-button"
          className={styles.search__button}
          type="submit"
          aria-label="Buscar por heróis"
          title="Buscar por heróis"
          disabled={disabled}>
          <img src="/assets/ic_busca.svg" alt="" width={29} height={29} />
        </button>

        <input
          className={styles.search__input}
          type="search"
          aria-label="Procure por heróis"
          title="Procure por heróis"
          placeholder="Procure por heróis"
          data-testid="search-input"
          name="search-hero"
          maxLength={120}
          disabled={disabled}
          defaultValue={inputValue}
        />
      </form>
    </div>
  );
}
