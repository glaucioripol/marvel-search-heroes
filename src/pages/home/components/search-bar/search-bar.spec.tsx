import {
  defaultRender,
  describe,
  expect,
  fireEvent,
  it,
  vi,
} from "@/__tests__";
import { SearchBar } from "../search-bar";

describe("SearchBar", () => {
  const sutSearchBar = (isDisabled = false) => {
    const handleSearch = vi.fn();

    const searchBarRendered = defaultRender(
      <SearchBar disabled={isDisabled} handleSearch={handleSearch} />,
    );

    return {
      ...searchBarRendered,
      handleSearchMock: handleSearch,
    };
  };

  it("renders correctly", () => {
    const { getByTestId } = sutSearchBar();
    expect(getByTestId("search-bar-wrapper")).toBeInTheDocument();
  });

  it("has search button and input field", () => {
    const { getByTestId } = sutSearchBar();

    expect(getByTestId("search-button")).toBeInTheDocument();
    expect(getByTestId("search-input")).toBeInTheDocument();
    expect(getByTestId("search-bar-wrapper")).toMatchSnapshot();
  });

  it("disables search button and input field when disabled prop is true", () => {
    const { getByTestId } = sutSearchBar(true);
    expect(getByTestId("search-button")).toBeDisabled();
    expect(getByTestId("search-input")).toBeDisabled();
    expect(getByTestId("search-bar-wrapper")).toMatchSnapshot();
  });

  it("calls handleSearch function when form is submitted", () => {
    const { getByTestId, handleSearchMock } = sutSearchBar();

    const mockedData = { value: "spider" };
    fireEvent.change(getByTestId("search-input"), {
      target: mockedData,
    });

    fireEvent.submit(getByTestId("search-form"));

    // will call handleSearchMock with "spider" as argument the field called search-hero
    expect(handleSearchMock).toHaveBeenCalledWith(
      expect.objectContaining({
        target: getByTestId("search-form"),
      }),
    );

    // check value of input
    expect(getByTestId("search-input")).toHaveValue("spider");

    expect(getByTestId("search-bar-wrapper")).toMatchSnapshot();
  });
});
