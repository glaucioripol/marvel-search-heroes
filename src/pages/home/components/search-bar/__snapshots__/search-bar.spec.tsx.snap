// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > calls handleSearch function when form is submitted 1`] = `
<div
  class="_search_04ea97"
  data-testid="search-bar-wrapper"
>
  <form
    class="_search__form_04ea97"
    data-testid="search-form"
  >
    <button
      aria-label="Buscar por heróis"
      class="_search__button_04ea97"
      data-testid="search-button"
      title="Buscar por heróis"
      type="submit"
    >
      <img
        alt="Ícone de uma lupa representando a busca."
        height="29"
        src="/assets/ic_busca.svg"
        width="29"
      />
    </button>
    <input
      aria-label="Procure por heróis"
      class="_search__input_04ea97"
      data-testid="search-input"
      maxlength="120"
      name="search-hero"
      placeholder="Procure por heróis"
      title="Procure por heróis"
      type="search"
      value=""
    />
  </form>
</div>
`;

exports[`SearchBar > disables search button and input field when disabled prop is true 1`] = `
<div
  class="_search_04ea97"
  data-testid="search-bar-wrapper"
>
  <form
    class="_search__form_04ea97"
    data-testid="search-form"
  >
    <button
      aria-label="Buscar por heróis"
      class="_search__button_04ea97"
      data-testid="search-button"
      disabled=""
      title="Buscar por heróis"
      type="submit"
    >
      <img
        alt="Ícone de uma lupa representando a busca."
        height="29"
        src="/assets/ic_busca.svg"
        width="29"
      />
    </button>
    <input
      aria-label="Procure por heróis"
      class="_search__input_04ea97"
      data-testid="search-input"
      disabled=""
      maxlength="120"
      name="search-hero"
      placeholder="Procure por heróis"
      title="Procure por heróis"
      type="search"
      value=""
    />
  </form>
</div>
`;

exports[`SearchBar > has search button and input field 1`] = `
<div
  class="_search_04ea97"
  data-testid="search-bar-wrapper"
>
  <form
    class="_search__form_04ea97"
    data-testid="search-form"
  >
    <button
      aria-label="Buscar por heróis"
      class="_search__button_04ea97"
      data-testid="search-button"
      title="Buscar por heróis"
      type="submit"
    >
      <img
        alt="Ícone de uma lupa representando a busca."
        height="29"
        src="/assets/ic_busca.svg"
        width="29"
      />
    </button>
    <input
      aria-label="Procure por heróis"
      class="_search__input_04ea97"
      data-testid="search-input"
      maxlength="120"
      name="search-hero"
      placeholder="Procure por heróis"
      title="Procure por heróis"
      type="search"
      value=""
    />
  </form>
</div>
`;
